# Notebook goals
First we're going to derive wins, wins in last 10 games, Pythagorean Wins and Run Differential from our scores dataframe. Then we'll filter out doubledeaders, and we'll explain why when we get to it.<br>

We actually went through this notebook twice, once for 2021 and again for 2022. When looking back at this notebook we'll see the code from whichever year was most recently processed, even though the variables still hvae "21" in their name.

In [140]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests
import unidecode
from datetime import datetime
from datetime import timedelta

In [141]:
def chatGPT(text):
  url = "https://api.openai.com/v1/completions"
  headers = {
  "Content-Type": "application/json",
  "Authorization": "Bearer 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",
  }
  data = { 
  "model": "text-davinci-003",
  "prompt": text,
  "max_tokens": 4000,
  "temperature": 1.0,
  }
  response = requests.post(url, headers=headers, json=data)
  output = response.json()["choices"][0]["text"]
  
  return print(output)

In [142]:
scores_df = pd.read_csv('Game_Scores_2021-22.csv')

In [143]:
scores_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4859 entries, 0 to 4858
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  4859 non-null   int64 
 1   Away_Team   4859 non-null   object
 2   Away_Score  4859 non-null   int64 
 3   Home_Team   4859 non-null   object
 4   Home_Score  4859 non-null   int64 
 5   Date        4859 non-null   object
dtypes: int64(3), object(3)
memory usage: 227.9+ KB


In [144]:
scores_df.drop(columns = ['Unnamed: 0'], inplace = True)

In [145]:
scores_df['Date'] = pd.to_datetime(scores_df['Date'])

In [146]:
scores_df.head()

Away_Team  Away_Score Home_Team  Home_Score       Date
0       CHW           3       LAA           4 2021-04-01
1       PIT           5       CHC           3 2021-04-01
2       STL          11       CIN           6 2021-04-01
3       LAD           5       COL           8 2021-04-01
4       CLE           2       DET           3 2021-04-01

In [147]:
scores_df.tail()

Away_Team  Away_Score Home_Team  Home_Score       Date
4854       LAA           2       OAK           3 2022-10-05
4855       STL           3       PIT           5 2022-10-05
4856       SFG           8       SDP           1 2022-10-05
4857       DET           4       SEA           5 2022-10-05
4858       NYY           2       TEX           4 2022-10-05

# Win variables
Let's add columns indicating whether the home or away team won the game.

In [1]:
scores_df['Home_Win'] = np.where(scores_df['Home_Score'] > scores_df['Away_Score'], 1, 0)
scores_df['Away_Win'] = np.where(scores_df['Away_Score'] > scores_df['Home_Score'], 1, 0)

NameError: name 'np' is not defined

In [246]:
scores_df.head()

Away_Team  Away_Score Home_Team  Home_Score       Date  Home_Win  Away_Win
0       CHW           3       LAA           4 2021-04-01         1         0
1       PIT           5       CHC           3 2021-04-01         0         1
2       STL          11       CIN           6 2021-04-01         0         1
3       LAD           5       COL           8 2021-04-01         1         0
4       CLE           2       DET           3 2021-04-01         1         0

In [247]:
scores_df.tail()

Away_Team  Away_Score Home_Team  Home_Score       Date  Home_Win  \
4854       LAA           2       OAK           3 2022-10-05         1   
4855       STL           3       PIT           5 2022-10-05         1   
4856       SFG           8       SDP           1 2022-10-05         0   
4857       DET           4       SEA           5 2022-10-05         1   
4858       NYY           2       TEX           4 2022-10-05         1   

      Away_Win  
4854         0  
4855         0  
4856         1  
4857         0  
4858         0

In [248]:
scores_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4859 entries, 0 to 4858
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Away_Team   4859 non-null   object        
 1   Away_Score  4859 non-null   int64         
 2   Home_Team   4859 non-null   object        
 3   Home_Score  4859 non-null   int64         
 4   Date        4859 non-null   datetime64[ns]
 5   Home_Win    4859 non-null   int32         
 6   Away_Win    4859 non-null   int32         
dtypes: datetime64[ns](1), int32(2), int64(2), object(2)
memory usage: 227.9+ KB


## Filtering scores_df by 2021 and 2022 seasons

In [249]:
scores_21 = scores_df[scores_df['Date'] > '2021-12-31']

In [250]:
scores_21.head()

Away_Team  Away_Score Home_Team  Home_Score       Date  Home_Win  \
2429       HOU           3       LAA           1 2022-04-07         0   
2430       SDP           2       ARI           4 2022-04-07         1   
2431       CIN           6       ATL           3 2022-04-07         0   
2432       MIL           4       CHC           5 2022-04-07         1   
2433       CLE           1       KCR           3 2022-04-07         1   

      Away_Win  
2429         1  
2430         0  
2431         1  
2432         0  
2433         0

In [251]:
scores_21.tail()

Away_Team  Away_Score Home_Team  Home_Score       Date  Home_Win  \
4854       LAA           2       OAK           3 2022-10-05         1   
4855       STL           3       PIT           5 2022-10-05         1   
4856       SFG           8       SDP           1 2022-10-05         0   
4857       DET           4       SEA           5 2022-10-05         1   
4858       NYY           2       TEX           4 2022-10-05         1   

      Away_Win  
4854         0  
4855         0  
4856         1  
4857         0  
4858         0

In [252]:
scores_21.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 2429 to 4858
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Away_Team   2430 non-null   object        
 1   Away_Score  2430 non-null   int64         
 2   Home_Team   2430 non-null   object        
 3   Home_Score  2430 non-null   int64         
 4   Date        2430 non-null   datetime64[ns]
 5   Home_Win    2430 non-null   int32         
 6   Away_Win    2430 non-null   int32         
dtypes: datetime64[ns](1), int32(2), int64(2), object(2)
memory usage: 132.9+ KB


# Inventory
**2021:** In a normal season, an MLB team plays 162 games, 81 at home and 81 on the road. We know the COL and ATL both played only 161 in 2021, because a few things were different that year because of COVID. Now we see that TOR has an extra road game. They couldn't play their home games in Canada until July 30 because of COVID. After a little digging we found that out a Blue Jays "home" game against the Angels (LAA) was postponed in April and made up in August at the Angels' home in Anaheim as part of a doubleheader. The Blue Jays were the "home" team for the first game of the doubleheader, meaning they had the last at-bat.<br>

So should the Jays actually be considered the home team in this game since they have the built-in home team advantage of having the last at-bat? Fortunately, we don't have to answer this question since we're going to filter out doubleheaders anyway. That's not the original reason we decided to take out doubleheaders, but in stumbling upon this situation during data cleaning it gives us another reason to do so.<br>

**2022:** COVID wasn't as much of a factor, but because a lockout delayed the start of the season by about a week, the A's and Tigers played a doubleheader in Detroit in which the A's were the home team for the first game. That meant that the A's were on the road for 82 games and the Tigers for 80, and vice versa. Again, just like we did for 2021, we'll filter these games out anyway since it's a doubleheader.

In [253]:
scores_21['Away_Team'].value_counts()

OAK    82
TBR    81
WSN    81
BOS    81
MIL    81
CLE    81
MIN    81
MIA    81
LAA    81
TEX    81
TOR    81
CHC    81
NYY    81
SEA    81
CHW    81
CIN    81
SDP    81
KCR    81
NYM    81
STL    81
PHI    81
LAD    81
COL    81
SFG    81
ATL    81
HOU    81
BAL    81
ARI    81
PIT    81
DET    80
Name: Away_Team, dtype: int64

# Explaining the 80s and 82s
We see above and below that TOR has 82 road games and 80 home games. That makes sense considering what we just talked about, as does LAA having 80 road games and 82 home games.<br>

As we said, the Rockies and Braves both played only 161 games in 2021, which means a scheduled game between those two teams wasn't played. This explains COL having 80 road games and 81 home games as well as ATL having 81 road games and 80 home games.

As far as 2022 goes, that's been explained in the previous markdown cell.

In [254]:
scores_21['Home_Team'].value_counts()

DET    82
TBR    81
WSN    81
BOS    81
MIL    81
LAA    81
MIN    81
MIA    81
CLE    81
TEX    81
TOR    81
CHC    81
NYY    81
SEA    81
CHW    81
CIN    81
SDP    81
KCR    81
NYM    81
STL    81
PHI    81
LAD    81
COL    81
SFG    81
ATL    81
HOU    81
BAL    81
ARI    81
PIT    81
OAK    80
Name: Home_Team, dtype: int64

Home Wins and Away Wins add up to 2429, the number of rows in the 2021 dataset. So that checks out.<br>

For 2022, the Home Wins and Away Wins add up to 2430.

In [255]:
scores_21['Home_Win'].sum()

1296

In [256]:
scores_21['Away_Win'].sum()

1134

# How do we keep track of each team's wins?
We'll start by trying to create a game_id column, since currently every row represents a game played.

In [159]:
chatGPT("How can I create a column that is a copy of the index?")



You can use the reset_index() method to add an additional column of the index as a new column in the DataFrame.

Example: 

import pandas as pd 

# Sample data 

data = [[1, 2, 3],[4, 5, 6],[7, 8, 9]]

df = pd.DataFrame(data, columns =['A','B','C']) 

# Add index column 

df = df.reset_index()

# Result 

df 
      index  A  B  C 
    0      0  1  2  3 
    1      1  4  5  6 
    2      2  7  8  9


In [257]:
scores_21 = scores_21.reset_index()

In [258]:
scores_21.head()

index Away_Team  Away_Score Home_Team  Home_Score       Date  Home_Win  \
0   2429       HOU           3       LAA           1 2022-04-07         0   
1   2430       SDP           2       ARI           4 2022-04-07         1   
2   2431       CIN           6       ATL           3 2022-04-07         0   
3   2432       MIL           4       CHC           5 2022-04-07         1   
4   2433       CLE           1       KCR           3 2022-04-07         1   

   Away_Win  
0         1  
1         0  
2         1  
3         0  
4         0

In [259]:
scores_21.rename(columns = {'index':'game_id'}, inplace = True)

In [260]:
scores_21.head()

game_id Away_Team  Away_Score Home_Team  Home_Score       Date  Home_Win  \
0     2429       HOU           3       LAA           1 2022-04-07         0   
1     2430       SDP           2       ARI           4 2022-04-07         1   
2     2431       CIN           6       ATL           3 2022-04-07         0   
3     2432       MIL           4       CHC           5 2022-04-07         1   
4     2433       CLE           1       KCR           3 2022-04-07         1   

   Away_Win  
0         1  
1         0  
2         1  
3         0  
4         0

# Home team indicator
Let's create a boolean to indicate home team. For now, we'll create one for each team. Obviously, Away_Host is False and Home_Host is True. But it seems like we need to set the variable now before splitting the dataframe.

In [1]:
scores_21['Away_Host'] = 0
scores_21['Home_Host'] = 1

NameError: name 'scores_21' is not defined

Now we'll split the dataframe.

In [263]:
away_21 = scores_21[['Date', 'game_id', 'Away_Team', 'Away_Score', 'Away_Win', 'Away_Host']]
home_21 = scores_21[['Date', 'game_id', 'Home_Team', 'Home_Score', 'Home_Win', 'Home_Host']]

In [264]:
away_21.head()

Date  game_id Away_Team  Away_Score  Away_Win  Away_Host
0 2022-04-07     2429       HOU           3         1          0
1 2022-04-07     2430       SDP           2         0          0
2 2022-04-07     2431       CIN           6         1          0
3 2022-04-07     2432       MIL           4         0          0
4 2022-04-07     2433       CLE           1         0          0

In [265]:
home_21.head()

Date  game_id Home_Team  Home_Score  Home_Win  Home_Host
0 2022-04-07     2429       LAA           1         0          1
1 2022-04-07     2430       ARI           4         1          1
2 2022-04-07     2431       ATL           3         0          1
3 2022-04-07     2432       CHC           5         1          1
4 2022-04-07     2433       KCR           3         1          1

In [266]:
away_21 = away_21.rename(columns = {'Away_Team': 'Team', 'Away_Score': 'Score', 'Away_Win': 'Win', 'Away_Host': 'Host'})
home_21 = home_21.rename(columns = {'Home_Team': 'Team', 'Home_Score': 'Score', 'Home_Win': 'Win', 'Home_Host': 'Host'})

In [267]:
away_21.head(10)

Date  game_id Team  Score  Win  Host
0 2022-04-07     2429  HOU      3    1     0
1 2022-04-07     2430  SDP      2    0     0
2 2022-04-07     2431  CIN      6    1     0
3 2022-04-07     2432  MIL      4    0     0
4 2022-04-07     2433  CLE      1    0     0
5 2022-04-07     2434  PIT      0    0     0
6 2022-04-07     2435  NYM      5    1     0
7 2022-04-08     2436  HOU     13    1     0
8 2022-04-08     2437  SDP      3    1     0
9 2022-04-08     2438  CIN      6    0     0

In [268]:
home_21.head()

Date  game_id Team  Score  Win  Host
0 2022-04-07     2429  LAA      1    0     1
1 2022-04-07     2430  ARI      4    1     1
2 2022-04-07     2431  ATL      3    0     1
3 2022-04-07     2432  CHC      5    1     1
4 2022-04-07     2433  KCR      3    1     1

# One dataframe, one team per row
How can we splice the home and away dataframes into one so that each team has its own row, and every game is represented by a pair of rows.

In [269]:
chatGPT("Can do pd.concat but then mix the rows of the dataframe and order the new dataframe by certain columns?")



Yes, it is possible to use pd.concat to mix the rows of a dataframe and then reorder it by certain columns.
This can be done by passing an additional argument 'sort' as part of the function which will specify the columns by which reordering should be done.

For example:

df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'], 
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                   index=[0, 1, 2, 3])

df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                   index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'],
                    'D': ['D8', 'D9', 'D10', 'D11']},
                   ind

In [270]:
wins_21 = pd.concat([away_21, home_21]).sort_values(['Date', 'game_id', 'Host'])

In [271]:
wins_21.head(50)

Date  game_id Team  Score  Win  Host
0  2022-04-07     2429  HOU      3    1     0
0  2022-04-07     2429  LAA      1    0     1
1  2022-04-07     2430  SDP      2    0     0
1  2022-04-07     2430  ARI      4    1     1
2  2022-04-07     2431  CIN      6    1     0
2  2022-04-07     2431  ATL      3    0     1
3  2022-04-07     2432  MIL      4    0     0
3  2022-04-07     2432  CHC      5    1     1
4  2022-04-07     2433  CLE      1    0     0
4  2022-04-07     2433  KCR      3    1     1
5  2022-04-07     2434  PIT      0    0     0
5  2022-04-07     2434  STL      9    1     1
6  2022-04-07     2435  NYM      5    1     0
6  2022-04-07     2435  WSN      1    0     1
7  2022-04-08     2436  HOU     13    1     0
7  2022-04-08     2436  LAA      6    0     1
8  2022-04-08     2437  SDP      3    1     0
8  2022-04-08     2437  ARI      0    0     1
9  2022-04-08     2438  CIN      6    0     0
9  2022-04-08     2438  ATL      7    1     1
10 2022-04-08     2439  LAD      5    1     0
10 2022-04-08     2439  COL      3    0     1
11 2022-04-08     2440  CHW      4    0     0
11 2022-04-08     2440  DET      5    1     1
12 2022-04-08     2441  SEA      2    1     0
12 2022-04-08     2441  MIN      1    0     1
13 2022-04-08     2442  BOS      5    0     0
13 2022-04-08     2442  NYY      6    1     1
14 2022-04-08     2443  OAK      5    0     0
14 2022-04-08     2443  PHI      9    1     1
15 2022-04-08     2444  MIA      5    0     0
15 2022-04-08     2444  SFG      6    1     1
16 2022-04-08     2445  BAL      1    0     0
16 2022-04-08     2445  TBR      2    1     1
17 2022-04-08     2446  TEX      8    0     0
17 2022-04-08     2446  TOR     10    1     1
18 2022-04-08     2447  NYM      7    1     0
18 2022-04-08     2447  WSN      3    0     1
19 2022-04-09     2448  HOU      0    0     0
19 2022-04-09     2448  LAA      2    1     1
20 2022-04-09     2449  SDP      5    1     0
20 2022-04-09     2449  ARI      2    0     1
21 2022-04-09     2450  CIN      1    0     0
21 2022-04-09     2450  ATL      2    1     1
22 2022-04-09     2451  MIL      0    0     0
22 2022-04-09     2451  CHC      9    1     1
23 2022-04-09     2452  LAD      2    0     0
23 2022-04-09     2452  COL      3    1     1
24 2022-04-09     2453  CHW      5    1     0
24 2022-04-09     2453  DET      2    0     1

In [272]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4860 entries, 0 to 2429
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   Date     4860 non-null   datetime64[ns]
 1   game_id  4860 non-null   int64         
 2   Team     4860 non-null   object        
 3   Score    4860 non-null   int64         
 4   Win      4860 non-null   int32         
 5   Host     4860 non-null   int64         
dtypes: datetime64[ns](1), int32(1), int64(3), object(1)
memory usage: 246.8+ KB


This confirms that the midpoint of the dataframe doesn't start again on April 1 (April 7 for 2022).

In [273]:
wins_21.iloc[2425:2435, :]

Date  game_id Team  Score  Win  Host
1212 2022-07-05     3641  OAK      5    1     1
1213 2022-07-05     3642  WSN      0    0     0
1213 2022-07-05     3642  PHI     11    1     1
1214 2022-07-05     3643  NYY      2    0     0
1214 2022-07-05     3643  PIT      5    1     1
1215 2022-07-05     3644  SEA      6    1     0
1215 2022-07-05     3644  SDP      2    0     1
1216 2022-07-06     3645  SFG      7    1     0
1216 2022-07-06     3645  ARI      5    0     1
1217 2022-07-06     3646  STL      0    0     0

All 162 except COL and ATL in 2021. All 162 in 2022. All good there.

In [275]:
wins_21['Team'].value_counts()

WSN    162
ARI    162
MIL    162
LAA    162
MIN    162
MIA    162
TOR    162
CHC    162
OAK    162
NYY    162
SDP    162
LAD    162
COL    162
SFG    162
ATL    162
HOU    162
BAL    162
DET    162
PIT    162
TEX    162
TBR    162
SEA    162
CHW    162
CIN    162
KCR    162
NYM    162
STL    162
PHI    162
CLE    162
BOS    162
Name: Team, dtype: int64

## Using groupby to get a cumulative Team Wins column

In [276]:
wins_21['Team_Wins'] = wins_21.groupby('Team')['Win'].cumsum()

# Tail of the tape
The best way to test our Team_Wins column is to look at the last rows and see if it matches the final team win totals of 2021.

In [277]:
wins_21.tail(50)

Date  game_id Team  Score  Win  Host  Team_Wins
2405 2022-10-04     4834  WSN      2    0     0         55
2405 2022-10-04     4834  NYM      4    1     1         99
2406 2022-10-04     4835  WSN      0    0     0         55
2406 2022-10-04     4835  NYM      8    1     1        100
2407 2022-10-04     4836  LAA      1    0     0         73
2407 2022-10-04     4836  OAK      2    1     1         59
2408 2022-10-04     4837  STL      8    1     0         93
2408 2022-10-04     4837  PIT      7    0     1         61
2409 2022-10-04     4838  SFG      2    0     0         80
2409 2022-10-04     4838  SDP      6    1     1         89
2410 2022-10-04     4839  DET      6    0     0         66
2410 2022-10-04     4839  SEA      7    1     1         88
2411 2022-10-04     4840  DET      6    0     0         66
2411 2022-10-04     4840  SEA      9    1     1         89
2412 2022-10-04     4841  NYY      5    1     0         99
2412 2022-10-04     4841  TEX      4    0     1         66
2413 2022-10-04     4842  NYY      2    0     0         99
2413 2022-10-04     4842  TEX      3    1     1         67
2414 2022-10-05     4843  TOR      4    0     0         91
2414 2022-10-05     4843  BAL      5    1     1         83
2415 2022-10-05     4844  TOR      5    1     0         92
2415 2022-10-05     4844  BAL      1    0     1         83
2416 2022-10-05     4845  TBR      3    0     0         86
2416 2022-10-05     4845  BOS      6    1     1         78
2417 2022-10-05     4846  MIN     10    1     0         78
2417 2022-10-05     4846  CHW      1    0     1         81
2418 2022-10-05     4847  CHC     15    1     0         74
2418 2022-10-05     4847  CIN      2    0     1         62
2419 2022-10-05     4848  KCR      2    0     0         65
2419 2022-10-05     4848  CLE      9    1     1         92
2420 2022-10-05     4849  PHI      2    0     0         87
2420 2022-10-05     4849  HOU      3    1     1        106
2421 2022-10-05     4850  COL      1    0     0         68
2421 2022-10-05     4850  LAD      6    1     1        111
2422 2022-10-05     4851  ATL      9    0     0        101
2422 2022-10-05     4851  MIA     12    1     1         69
2423 2022-10-05     4852  ARI      4    1     0         74
2423 2022-10-05     4852  MIL      2    0     1         86
2424 2022-10-05     4853  WSN      2    0     0         55
2424 2022-10-05     4853  NYM      9    1     1        101
2425 2022-10-05     4854  LAA      2    0     0         73
2425 2022-10-05     4854  OAK      3    1     1         60
2426 2022-10-05     4855  STL      3    0     0         93
2426 2022-10-05     4855  PIT      5    1     1         62
2427 2022-10-05     4856  SFG      8    1     0         81
2427 2022-10-05     4856  SDP      1    0     1         89
2428 2022-10-05     4857  DET      4    0     0         66
2428 2022-10-05     4857  SEA      5    1     1         90
2429 2022-10-05     4858  NYY      2    0     0         99
2429 2022-10-05     4858  TEX      4    1     1         68

# Last 10
Now we need a wins-in-last-10-games variable. This for loop creates a temporary dataframe for each team and subtracts Team_Wins from 10 rows before from Team_Wins in the current row of the loop. We need to reset the index of each temp_df so we can use the index when we use .loc.<br>

Then we concat the list of temp_dfs and sort by Date and game_id.

In [278]:
list_of_dfs = []
for team in wins_21['Team'].unique():
    temp_df = wins_21[wins_21['Team'] == team]
    temp_df = temp_df.reset_index(drop=True)
    temp_df['Wins_L10'] = 0
    for i in range(10):
        temp_df.loc[i, 'Wins_L10'] = temp_df.loc[i, 'Team_Wins']
    for j in range(10, (len(temp_df))):
        temp_df.loc[j, 'Wins_L10'] = temp_df.loc[j, 'Team_Wins'] - temp_df.loc[j-10, 'Team_Wins']
    list_of_dfs.append(temp_df)
        

In [279]:
wins_21 = pd.concat(list_of_dfs).sort_values(['Date', 'game_id', 'Host'])

# Last 30
Let's try the same for wins in last 30 games.

In [280]:
list_of_dfs = []
for team in wins_21['Team'].unique():
    temp_df = wins_21[wins_21['Team'] == team]
    temp_df = temp_df.reset_index(drop=True)
    temp_df['Wins_L30'] = 0
    for i in range(30):
        temp_df.loc[i, 'Wins_L30'] = temp_df.loc[i, 'Team_Wins']
    for j in range(30, (len(temp_df))):
        temp_df.loc[j, 'Wins_L30'] = temp_df.loc[j, 'Team_Wins'] - temp_df.loc[j-30, 'Team_Wins']
    list_of_dfs.append(temp_df)

In [281]:
wins_21 = pd.concat(list_of_dfs).sort_values(['Date', 'game_id'])

In [282]:
wins_21.head(50)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  Wins_L30
0 2022-04-07     2429  HOU      3    1     0          1         1         1
0 2022-04-07     2429  LAA      1    0     1          0         0         0
0 2022-04-07     2430  SDP      2    0     0          0         0         0
0 2022-04-07     2430  ARI      4    1     1          1         1         1
0 2022-04-07     2431  CIN      6    1     0          1         1         1
0 2022-04-07     2431  ATL      3    0     1          0         0         0
0 2022-04-07     2432  MIL      4    0     0          0         0         0
0 2022-04-07     2432  CHC      5    1     1          1         1         1
0 2022-04-07     2433  CLE      1    0     0          0         0         0
0 2022-04-07     2433  KCR      3    1     1          1         1         1
0 2022-04-07     2434  PIT      0    0     0          0         0         0
0 2022-04-07     2434  STL      9    1     1          1         1         1
0 2022-04-07     2435  NYM      5    1     0          1         1         1
0 2022-04-07     2435  WSN      1    0     1          0         0         0
1 2022-04-08     2436  HOU     13    1     0          2         2         2
1 2022-04-08     2436  LAA      6    0     1          0         0         0
1 2022-04-08     2437  SDP      3    1     0          1         1         1
1 2022-04-08     2437  ARI      0    0     1          1         1         1
1 2022-04-08     2438  CIN      6    0     0          1         1         1
1 2022-04-08     2438  ATL      7    1     1          1         1         1
0 2022-04-08     2439  LAD      5    1     0          1         1         1
0 2022-04-08     2439  COL      3    0     1          0         0         0
0 2022-04-08     2440  CHW      4    0     0          0         0         0
0 2022-04-08     2440  DET      5    1     1          1         1         1
0 2022-04-08     2441  SEA      2    1     0          1         1         1
0 2022-04-08     2441  MIN      1    0     1          0         0         0
0 2022-04-08     2442  BOS      5    0     0          0         0         0
0 2022-04-08     2442  NYY      6    1     1          1         1         1
0 2022-04-08     2443  OAK      5    0     0          0         0         0
0 2022-04-08     2443  PHI      9    1     1          1         1         1
0 2022-04-08     2444  MIA      5    0     0          0         0         0
0 2022-04-08     2444  SFG      6    1     1          1         1         1
0 2022-04-08     2445  BAL      1    0     0          0         0         0
0 2022-04-08     2445  TBR      2    1     1          1         1         1
0 2022-04-08     2446  TEX      8    0     0          0         0         0
0 2022-04-08     2446  TOR     10    1     1          1         1         1
1 2022-04-08     2447  NYM      7    1     0          2         2         2
1 2022-04-08     2447  WSN      3    0     1          0         0         0
2 2022-04-09     2448  HOU      0    0     0          2         2         2
2 2022-04-09     2448  LAA      2    1     1          1         1         1
2 2022-04-09     2449  SDP      5    1     0          2         2         2
2 2022-04-09     2449  ARI      2    0     1          1         1         1
2 2022-04-09     2450  CIN      1    0     0          1         1         1
2 2022-04-09     2450  ATL      2    1     1          2         2         2
1 2022-04-09     2451  MIL      0    0     0          0         0         0
1 2022-04-09     2451  CHC      9    1     1          2         2         2
1 2022-04-09     2452  LAD      2    0     0          1         1         1
1 2022-04-09     2452  COL      3    1     1          1         1         1
1 2022-04-09     2453  CHW      5    1     0          1         1         1
1 2022-04-09     2453  DET      2    0     1          1         1         1

In [283]:
wins_21['Wins_L10'].value_counts()

5     947
4     944
6     878
3     641
7     554
2     330
8     284
1     140
9      88
0      36
10     18
Name: Wins_L10, dtype: int64

In [284]:
wins_21['Wins_L30'].value_counts()

13    467
14    442
16    394
15    388
12    374
17    320
11    309
10    273
18    267
19    234
20    201
9     186
21    165
8     125
7     101
22    100
6      90
2      73
4      65
3      64
23     57
1      52
5      51
0      27
24     22
25      8
26      5
Name: Wins_L30, dtype: int64

# Pythagorean wins
Now let's add a Pythagorean Wins variable. According to <a href='https://www.mlb.com/glossary/advanced-stats/pythagorean-winning-percentage'>MLB.com</a>, this is a formula created by noted baseball statistician Bill James to measure how many games a team "should have" won, based on runs scored and runs allowed.<br>

**Formula:** Runs Scored to the 1.83 power/((Runs Scored to the 1.83 power) + (Runs Allowed to the 1.83 power))<br>

We'll start by creating a variable that counts each team's games played, then cumulative Runs_Scored and Runs_Allowed variables.<br>

In [285]:
wins_21 = wins_21.reset_index(drop=True)

In [286]:
wins_21['Games_Played'] = wins_21.groupby('Team').cumcount() + 1

In [287]:
wins_21.tail(10)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  \
4850 2022-10-05     4854  LAA      2    0     0         73         7   
4851 2022-10-05     4854  OAK      3    1     1         60         4   
4852 2022-10-05     4855  PIT      5    1     1         62         6   
4853 2022-10-05     4855  STL      3    0     0         93         4   
4854 2022-10-05     4856  SDP      1    0     1         89         5   
4855 2022-10-05     4856  SFG      8    1     0         81         7   
4856 2022-10-05     4857  DET      4    0     0         66         6   
4857 2022-10-05     4857  SEA      5    1     1         90         7   
4858 2022-10-05     4858  NYY      2    0     0         99         5   
4859 2022-10-05     4858  TEX      4    1     1         68         3   

      Wins_L30  Games_Played  
4850        16           162  
4851        11           162  
4852        13           162  
4853        16           162  
4854        16           162  
4855        17           162  
4856        15           162  
4857        16           162  
4858        20           162  
4859        10           162

In [288]:
wins_21['Opp_Score'] = 0

In [289]:
for i in range(len(wins_21)):
    if i % 2 == 0:
        wins_21.loc[i, 'Opp_Score'] = wins_21.loc[i+1, 'Score']
    else:
        wins_21.loc[i, 'Opp_Score'] = wins_21.loc[i-1, 'Score']
            
    

In [290]:
wins_21.head(10)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  Wins_L30  \
0 2022-04-07     2429  HOU      3    1     0          1         1         1   
1 2022-04-07     2429  LAA      1    0     1          0         0         0   
2 2022-04-07     2430  SDP      2    0     0          0         0         0   
3 2022-04-07     2430  ARI      4    1     1          1         1         1   
4 2022-04-07     2431  CIN      6    1     0          1         1         1   
5 2022-04-07     2431  ATL      3    0     1          0         0         0   
6 2022-04-07     2432  MIL      4    0     0          0         0         0   
7 2022-04-07     2432  CHC      5    1     1          1         1         1   
8 2022-04-07     2433  CLE      1    0     0          0         0         0   
9 2022-04-07     2433  KCR      3    1     1          1         1         1   

   Games_Played  Opp_Score  
0             1          1  
1             1          3  
2             1          4  
3             1          2  
4             1          3  
5             1          6  
6             1          5  
7             1          4  
8             1          3  
9             1          1

In [291]:
wins_21['Runs_Scored'] = wins_21.groupby('Team')['Score'].cumsum()

In [292]:
wins_21.tail(10)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  \
4850 2022-10-05     4854  LAA      2    0     0         73         7   
4851 2022-10-05     4854  OAK      3    1     1         60         4   
4852 2022-10-05     4855  PIT      5    1     1         62         6   
4853 2022-10-05     4855  STL      3    0     0         93         4   
4854 2022-10-05     4856  SDP      1    0     1         89         5   
4855 2022-10-05     4856  SFG      8    1     0         81         7   
4856 2022-10-05     4857  DET      4    0     0         66         6   
4857 2022-10-05     4857  SEA      5    1     1         90         7   
4858 2022-10-05     4858  NYY      2    0     0         99         5   
4859 2022-10-05     4858  TEX      4    1     1         68         3   

      Wins_L30  Games_Played  Opp_Score  Runs_Scored  
4850        16           162          3          623  
4851        11           162          2          568  
4852        13           162          3          591  
4853        16           162          5          772  
4854        16           162          8          705  
4855        17           162          1          716  
4856        15           162          5          557  
4857        16           162          4          690  
4858        20           162          4          807  
4859        10           162          2          707

In [293]:
wins_21['Runs_Allowed'] = wins_21.groupby('Team')['Opp_Score'].cumsum()

In [294]:
wins_21.tail(10)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  \
4850 2022-10-05     4854  LAA      2    0     0         73         7   
4851 2022-10-05     4854  OAK      3    1     1         60         4   
4852 2022-10-05     4855  PIT      5    1     1         62         6   
4853 2022-10-05     4855  STL      3    0     0         93         4   
4854 2022-10-05     4856  SDP      1    0     1         89         5   
4855 2022-10-05     4856  SFG      8    1     0         81         7   
4856 2022-10-05     4857  DET      4    0     0         66         6   
4857 2022-10-05     4857  SEA      5    1     1         90         7   
4858 2022-10-05     4858  NYY      2    0     0         99         5   
4859 2022-10-05     4858  TEX      4    1     1         68         3   

      Wins_L30  Games_Played  Opp_Score  Runs_Scored  Runs_Allowed  
4850        16           162          3          623           668  
4851        11           162          2          568           770  
4852        13           162          3          591           817  
4853        16           162          5          772           637  
4854        16           162          8          705           660  
4855        17           162          1          716           697  
4856        15           162          5          557           713  
4857        16           162          4          690           623  
4858        20           162          4          807           567  
4859        10           162          2          707           743

In [295]:
wins_21['Py_Wins'] = np.round(wins_21['Runs_Scored']**1.83/((wins_21['Runs_Scored']**1.83)+(wins_21['Runs_Allowed']**1.83))\
*wins_21['Games_Played'], 0)

In [296]:
wins_21.iloc[1000:1005, :]

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  \
1000 2022-05-14     2929  TBR      1    0     1         20         6   
1001 2022-05-14     2929  TOR      5    1     0         18         3   
1002 2022-05-14     2930  BOS     11    1     0         13         4   
1003 2022-05-14     2930  TEX      3    0     1         13         5   
1004 2022-05-14     2931  HOU      6    0     0         22         9   

      Wins_L30  Games_Played  Opp_Score  Runs_Scored  Runs_Allowed  Py_Wins  
1000        17            34          5          143           147     17.0  
1001        15            34          1          130           142     16.0  
1002        12            33          3          125           131     16.0  
1003        13            32         11          124           143     14.0  
1004        19            34         13          143           111     21.0

Now our runs-differential feature

In [297]:
wins_21['Run_Diff'] = wins_21['Runs_Scored'] - wins_21['Runs_Allowed']

In [298]:
wins_21.head(10)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  Wins_L30  \
0 2022-04-07     2429  HOU      3    1     0          1         1         1   
1 2022-04-07     2429  LAA      1    0     1          0         0         0   
2 2022-04-07     2430  SDP      2    0     0          0         0         0   
3 2022-04-07     2430  ARI      4    1     1          1         1         1   
4 2022-04-07     2431  CIN      6    1     0          1         1         1   
5 2022-04-07     2431  ATL      3    0     1          0         0         0   
6 2022-04-07     2432  MIL      4    0     0          0         0         0   
7 2022-04-07     2432  CHC      5    1     1          1         1         1   
8 2022-04-07     2433  CLE      1    0     0          0         0         0   
9 2022-04-07     2433  KCR      3    1     1          1         1         1   

   Games_Played  Opp_Score  Runs_Scored  Runs_Allowed  Py_Wins  Run_Diff  
0             1          1            3             1      1.0         2  
1             1          3            1             3      0.0        -2  
2             1          4            2             4      0.0        -2  
3             1          2            4             2      1.0         2  
4             1          3            6             3      1.0         3  
5             1          6            3             6      0.0        -3  
6             1          5            4             5      0.0        -1  
7             1          4            5             4      1.0         1  
8             1          3            1             3      0.0        -2  
9             1          1            3             1      1.0         2

In [299]:
#NEXT
#Shift the rows forward one so that all the win variables represent the data going into that day's game.

In [300]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4860 entries, 0 to 4859
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          4860 non-null   datetime64[ns]
 1   game_id       4860 non-null   int64         
 2   Team          4860 non-null   object        
 3   Score         4860 non-null   int64         
 4   Win           4860 non-null   int32         
 5   Host          4860 non-null   int64         
 6   Team_Wins     4860 non-null   int32         
 7   Wins_L10      4860 non-null   int64         
 8   Wins_L30      4860 non-null   int64         
 9   Games_Played  4860 non-null   int64         
 10  Opp_Score     4860 non-null   int64         
 11  Runs_Scored   4860 non-null   int64         
 12  Runs_Allowed  4860 non-null   int64         
 13  Py_Wins       4860 non-null   float64       
 14  Run_Diff      4860 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int3

# Dropping a few columns
We used these colums to derive features and no longer need them.

In [301]:
wins_21 = wins_21.drop(columns = ['Games_Played', 'Opp_Score', 'Runs_Scored', 'Runs_Allowed'])

In [302]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4860 entries, 0 to 4859
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       4860 non-null   datetime64[ns]
 1   game_id    4860 non-null   int64         
 2   Team       4860 non-null   object        
 3   Score      4860 non-null   int64         
 4   Win        4860 non-null   int32         
 5   Host       4860 non-null   int64         
 6   Team_Wins  4860 non-null   int32         
 7   Wins_L10   4860 non-null   int64         
 8   Wins_L30   4860 non-null   int64         
 9   Py_Wins    4860 non-null   float64       
 10  Run_Diff   4860 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int32(2), int64(6), object(1)
memory usage: 379.8+ KB


# Columns to shift
We'll create a list of columns to shift. Then once again we'll loop through all 30 teams and shift these columns down one so that we have this data **going into** the next game so that it's predictive.

In [303]:
cols_to_shift = ['Team_Wins', 'Wins_L10', 'Wins_L30', 'Run_Diff', 'Py_Wins']

In [304]:
chatGPT("How do I shift the values in a column down one in pandas?")



This can be done with the `shift()` method of the pandas DataFrame. For example: 

```
df['column_name'] = df['column_name'].shift(1)
```

This will shift all values in the `column_name` column down one row.


In [305]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4860 entries, 0 to 4859
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       4860 non-null   datetime64[ns]
 1   game_id    4860 non-null   int64         
 2   Team       4860 non-null   object        
 3   Score      4860 non-null   int64         
 4   Win        4860 non-null   int32         
 5   Host       4860 non-null   int64         
 6   Team_Wins  4860 non-null   int32         
 7   Wins_L10   4860 non-null   int64         
 8   Wins_L30   4860 non-null   int64         
 9   Py_Wins    4860 non-null   float64       
 10  Run_Diff   4860 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int32(2), int64(6), object(1)
memory usage: 379.8+ KB


In [306]:
list_of_dfs = []
for team in wins_21['Team'].unique():
    temp_df = wins_21[wins_21['Team'] == team]
    temp_df = temp_df.reset_index(drop=True)
    for col in cols_to_shift:
        temp_df[col] = temp_df[col].shift(periods = 1)
    list_of_dfs.append(temp_df)

In [307]:
wins_21 = pd.concat(list_of_dfs).sort_values(by = ['Date', 'game_id', 'Host'])

In [310]:
wins_21.tail(30)

Date  game_id Team  Score  Win  Host  Team_Wins  Wins_L10  Wins_L30  \
161 2022-10-05     4844  TOR      5    1     0       91.0       7.0      19.0   
161 2022-10-05     4844  BAL      1    0     1       83.0       4.0      13.0   
161 2022-10-05     4845  TBR      3    0     0       86.0       2.0      12.0   
161 2022-10-05     4845  BOS      6    1     1       77.0       5.0      14.0   
161 2022-10-05     4846  MIN     10    1     0       77.0       4.0      10.0   
161 2022-10-05     4846  CHW      1    0     1       81.0       5.0      16.0   
161 2022-10-05     4847  CHC     15    1     0       73.0       7.0      17.0   
161 2022-10-05     4847  CIN      2    0     1       62.0       3.0      10.0   
161 2022-10-05     4848  KCR      2    0     0       65.0       3.0      12.0   
161 2022-10-05     4848  CLE      9    1     1       91.0       7.0      23.0   
161 2022-10-05     4849  PHI      2    0     0       87.0       4.0      14.0   
161 2022-10-05     4849  HOU      3    1     1      105.0       6.0      21.0   
161 2022-10-05     4850  COL      1    0     0       68.0       3.0      12.0   
161 2022-10-05     4850  LAD      6    1     1      110.0       6.0      20.0   
161 2022-10-05     4851  ATL      9    0     0      101.0       8.0      21.0   
161 2022-10-05     4851  MIA     12    1     1       68.0       6.0      13.0   
161 2022-10-05     4852  ARI      4    1     0       73.0       3.0      10.0   
161 2022-10-05     4852  MIL      2    0     1       86.0       5.0      17.0   
161 2022-10-05     4853  WSN      2    0     0       55.0       3.0      10.0   
161 2022-10-05     4853  NYM      9    1     1      100.0       5.0      17.0   
161 2022-10-05     4854  LAA      2    0     0       73.0       7.0      16.0   
161 2022-10-05     4854  OAK      3    1     1       59.0       4.0      10.0   
161 2022-10-05     4855  STL      3    0     0       93.0       5.0      17.0   
161 2022-10-05     4855  PIT      5    1     1       61.0       6.0      12.0   
161 2022-10-05     4856  SFG      8    1     0       80.0       6.0      17.0   
161 2022-10-05     4856  SDP      1    0     1       89.0       6.0      17.0   
161 2022-10-05     4857  DET      4    0     0       66.0       7.0      16.0   
161 2022-10-05     4857  SEA      5    1     1       89.0       6.0      16.0   
161 2022-10-05     4858  NYY      2    0     0       99.0       6.0      20.0   
161 2022-10-05     4858  TEX      4    1     1       67.0       2.0       9.0   

     Py_Wins  Run_Diff  
161     90.0      92.0  
161     79.0     -10.0  
161     87.0      55.0  
161     75.0     -55.0  
161     81.0       3.0  
161     78.0     -22.0  
161     71.0     -87.0  
161     65.0    -154.0  
161     64.0    -163.0  
161     87.0      57.0  
161     87.0      63.0  
161    106.0     218.0  
161     65.0    -170.0  
161    115.0     329.0  
161    100.0     183.0  
161     70.0     -93.0  
161     76.0     -40.0  
161     85.0      39.0  
161     56.0    -245.0  
161     97.0     159.0  
161     75.0     -44.0  
161     58.0    -203.0  
161     95.0     137.0  
161     57.0    -228.0  
161     82.0      12.0  
161     86.0      52.0  
161     63.0    -155.0  
161     88.0      66.0  
161    106.0     242.0  
161     77.0     -38.0

As expected, we have 30 missing values for each of the features we shifted.

In [311]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4860 entries, 0 to 161
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       4860 non-null   datetime64[ns]
 1   game_id    4860 non-null   int64         
 2   Team       4860 non-null   object        
 3   Score      4860 non-null   int64         
 4   Win        4860 non-null   int32         
 5   Host       4860 non-null   int64         
 6   Team_Wins  4830 non-null   float64       
 7   Wins_L10   4830 non-null   float64       
 8   Wins_L30   4830 non-null   float64       
 9   Py_Wins    4830 non-null   float64       
 10  Run_Diff   4830 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int32(1), int64(3), object(1)
memory usage: 436.6+ KB


# Back to the wide dataframe
Now we need to re-assemble the original dataframe so that each row represents a single game. We can start by creating two new dataframes based on the Host value.

In [312]:
wins_21_A = wins_21[wins_21['Host'] == 0]
wins_21_H = wins_21[wins_21['Host'] == 1]

In [313]:
wins_21_A.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 161
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       2430 non-null   datetime64[ns]
 1   game_id    2430 non-null   int64         
 2   Team       2430 non-null   object        
 3   Score      2430 non-null   int64         
 4   Win        2430 non-null   int32         
 5   Host       2430 non-null   int64         
 6   Team_Wins  2415 non-null   float64       
 7   Wins_L10   2415 non-null   float64       
 8   Wins_L30   2415 non-null   float64       
 9   Py_Wins    2415 non-null   float64       
 10  Run_Diff   2415 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int32(1), int64(3), object(1)
memory usage: 218.3+ KB


In [314]:
wins_21_H.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 161
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       2430 non-null   datetime64[ns]
 1   game_id    2430 non-null   int64         
 2   Team       2430 non-null   object        
 3   Score      2430 non-null   int64         
 4   Win        2430 non-null   int32         
 5   Host       2430 non-null   int64         
 6   Team_Wins  2415 non-null   float64       
 7   Wins_L10   2415 non-null   float64       
 8   Wins_L30   2415 non-null   float64       
 9   Py_Wins    2415 non-null   float64       
 10  Run_Diff   2415 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int32(1), int64(3), object(1)
memory usage: 218.3+ KB


# Renaming columns
Now we'll rename the columns of each dataframe so that the away-team data has an 'A' prefix and the home-team data has an 'H' prefix. Then we'll drop the 'Host' column from each since the column names now indicate home or away.

In [315]:
wins_21_A = wins_21_A.rename(columns = {'Team': 'A_Team', 'Score': 'A_Score', 'Win': 'A_Win', 'Team_Wins': 'A_Team_Wins',\
                                        'Wins_L10': 'A_Wins_L10', 'Wins_L30': 'A_Wins_L30', 'Py_Wins': 'A_Py_Wins',\
                                       'Run_Diff': 'A_Run_Diff'})

In [316]:
wins_21_A = wins_21_A.drop(columns = 'Host')

In [317]:
wins_21_A.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 161
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         2430 non-null   datetime64[ns]
 1   game_id      2430 non-null   int64         
 2   A_Team       2430 non-null   object        
 3   A_Score      2430 non-null   int64         
 4   A_Win        2430 non-null   int32         
 5   A_Team_Wins  2415 non-null   float64       
 6   A_Wins_L10   2415 non-null   float64       
 7   A_Wins_L30   2415 non-null   float64       
 8   A_Py_Wins    2415 non-null   float64       
 9   A_Run_Diff   2415 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int32(1), int64(2), object(1)
memory usage: 199.3+ KB


In [318]:
wins_21_H = wins_21_H.rename(columns = {'Team': 'H_Team', 'Score': 'H_Score', 'Win': 'H_Win', 'Team_Wins': 'H_Team_Wins',\
                                        'Wins_L10': 'H_Wins_L10', 'Wins_L30': 'H_Wins_L30', 'Py_Wins': 'H_Py_Wins',\
                                       'Run_Diff': 'H_Run_Diff'})

In [319]:
wins_21_H = wins_21_H.drop(columns = 'Host')

In [320]:
wins_21_H.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 161
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         2430 non-null   datetime64[ns]
 1   game_id      2430 non-null   int64         
 2   H_Team       2430 non-null   object        
 3   H_Score      2430 non-null   int64         
 4   H_Win        2430 non-null   int32         
 5   H_Team_Wins  2415 non-null   float64       
 6   H_Wins_L10   2415 non-null   float64       
 7   H_Wins_L30   2415 non-null   float64       
 8   H_Py_Wins    2415 non-null   float64       
 9   H_Run_Diff   2415 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int32(1), int64(2), object(1)
memory usage: 199.3+ KB


In [321]:
wins_21 = pd.merge(wins_21_A, wins_21_H, on = ['Date', 'game_id'], how = 'left')

In [322]:
wins_21.head(30)

Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
0  2022-04-07     2429    HOU        3      1          NaN         NaN   
1  2022-04-07     2430    SDP        2      0          NaN         NaN   
2  2022-04-07     2431    CIN        6      1          NaN         NaN   
3  2022-04-07     2432    MIL        4      0          NaN         NaN   
4  2022-04-07     2433    CLE        1      0          NaN         NaN   
5  2022-04-07     2434    PIT        0      0          NaN         NaN   
6  2022-04-07     2435    NYM        5      1          NaN         NaN   
7  2022-04-08     2436    HOU       13      1          1.0         1.0   
8  2022-04-08     2437    SDP        3      1          0.0         0.0   
9  2022-04-08     2438    CIN        6      0          1.0         1.0   
10 2022-04-08     2439    LAD        5      1          NaN         NaN   
11 2022-04-08     2440    CHW        4      0          NaN         NaN   
12 2022-04-08     2441    SEA        2      1          NaN         NaN   
13 2022-04-08     2442    BOS        5      0          NaN         NaN   
14 2022-04-08     2443    OAK        5      0          NaN         NaN   
15 2022-04-08     2444    MIA        5      0          NaN         NaN   
16 2022-04-08     2445    BAL        1      0          NaN         NaN   
17 2022-04-08     2446    TEX        8      0          NaN         NaN   
18 2022-04-08     2447    NYM        7      1          1.0         1.0   
19 2022-04-09     2448    HOU        0      0          2.0         2.0   
20 2022-04-09     2449    SDP        5      1          1.0         1.0   
21 2022-04-09     2450    CIN        1      0          1.0         1.0   
22 2022-04-09     2451    MIL        0      0          0.0         0.0   
23 2022-04-09     2452    LAD        2      0          1.0         1.0   
24 2022-04-09     2453    CHW        5      1          0.0         0.0   
25 2022-04-09     2454    CLE        0      0          0.0         0.0   
26 2022-04-09     2455    SEA        4      1          1.0         1.0   
27 2022-04-09     2456    BOS        2      0          0.0         0.0   
28 2022-04-09     2457    OAK        2      0          0.0         0.0   
29 2022-04-09     2458    MIA        2      1          0.0         0.0   

    A_Wins_L30  A_Py_Wins  A_Run_Diff H_Team  H_Score  H_Win  H_Team_Wins  \
0          NaN        NaN         NaN    LAA        1      0          NaN   
1          NaN        NaN         NaN    ARI        4      1          NaN   
2          NaN        NaN         NaN    ATL        3      0          NaN   
3          NaN        NaN         NaN    CHC        5      1          NaN   
4          NaN        NaN         NaN    KCR        3      1          NaN   
5          NaN        NaN         NaN    STL        9      1          NaN   
6          NaN        NaN         NaN    WSN        1      0          NaN   
7          1.0        1.0         2.0    LAA        6      0          0.0   
8          0.0        0.0        -2.0    ARI        0      0          1.0   
9          1.0        1.0         3.0    ATL        7      1          0.0   
10         NaN        NaN         NaN    COL        3      0          NaN   
11         NaN        NaN         NaN    DET        5      1          NaN   
12         NaN        NaN         NaN    MIN        1      0          NaN   
13         NaN        NaN         NaN    NYY        6      1          NaN   
14         NaN        NaN         NaN    PHI        9      1          NaN   
15         NaN        NaN         NaN    SFG        6      1          NaN   
16         NaN        NaN         NaN    TBR        2      1          NaN   
17         NaN        NaN         NaN    TOR       10      1          NaN   
18         1.0        1.0         4.0    WSN        3      0          0.0   
19         2.0        2.0         9.0    LAA        2      1          0.0   
20         1.0        1.0         1.0    ARI        2      0          1.0   
21         1.0      

In [323]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 2429
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         2430 non-null   datetime64[ns]
 1   game_id      2430 non-null   int64         
 2   A_Team       2430 non-null   object        
 3   A_Score      2430 non-null   int64         
 4   A_Win        2430 non-null   int32         
 5   A_Team_Wins  2415 non-null   float64       
 6   A_Wins_L10   2415 non-null   float64       
 7   A_Wins_L30   2415 non-null   float64       
 8   A_Py_Wins    2415 non-null   float64       
 9   A_Run_Diff   2415 non-null   float64       
 10  H_Team       2430 non-null   object        
 11  H_Score      2430 non-null   int64         
 12  H_Win        2430 non-null   int32         
 13  H_Team_Wins  2415 non-null   float64       
 14  H_Wins_L10   2415 non-null   float64       
 15  H_Wins_L30   2415 non-null   float64       
 16  H_Py_W

In [324]:
wins_21.iloc[1477, :]

Date           2022-07-28 00:00:00
game_id                       3906
A_Team                         TBR
A_Score                          0
A_Win                            0
A_Team_Wins                     53
A_Wins_L10                       5
A_Wins_L30                      16
A_Py_Wins                       52
A_Run_Diff                      24
H_Team                         BAL
H_Score                          3
H_Win                            1
H_Team_Wins                     49
H_Wins_L10                       5
H_Wins_L30                      19
H_Py_Wins                       48
H_Run_Diff                      -7
Name: 1477, dtype: object

# Doubleheaders
Now we get to doubleheaders. When it comes to doubleheaders, we'll have two scores and two starting pitchers on the same date for the same team. How can we be sure we're matching the pitcher to the game that he started? We'll go nuts trying to figure out a way, so let's just save time by dropping all data from doubleheaders.<br>

We'll start by grouping the scores dataframe by date, home team and away team and finding the rows with two home scores and away scores.

In [325]:
doubleheaders = wins_21.groupby(['Date', 'A_Team', 'H_Team']).count()

In [326]:
doubleheaders = doubleheaders.sort_values(by = 'A_Score', ascending = False)

In [327]:
doubleheaders.head()

game_id  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
Date       A_Team H_Team                                                     
2022-08-02 COL    SDP           2        2      2            2           2   
2022-05-04 SDP    CLE           2        2      2            2           2   
2022-09-17 CIN    STL           2        2      2            2           2   
2022-09-08 SFG    MIL           2        2      2            2           2   
2022-07-02 NYY    CLE           2        2      2            2           2   

                          A_Wins_L30  A_Py_Wins  A_Run_Diff  H_Score  H_Win  \
Date       A_Team H_Team                                                      
2022-08-02 COL    SDP              2          2           2        2      2   
2022-05-04 SDP    CLE              2          2           2        2      2   
2022-09-17 CIN    STL              2          2           2        2      2   
2022-09-08 SFG    MIL              2          2           2        2      2   
2022-07-02 NYY    CLE              2          2           2        2      2   

                          H_Team_Wins  H_Wins_L10  H_Wins_L30  H_Py_Wins  \
Date       A_Team H_Team                                                   
2022-08-02 COL    SDP               2           2           2          2   
2022-05-04 SDP    CLE               2           2           2          2   
2022-09-17 CIN    STL               2           2           2          2   
2022-09-08 SFG    MIL               2           2           2          2   
2022-07-02 NYY    CLE               2           2           2          2   

                          H_Run_Diff  
Date       A_Team H_Team              
2022-08-02 COL    SDP              2  
2022-05-04 SDP    CLE              2  
2022-09-17 CIN    STL              2  
2022-09-08 SFG    MIL              2  
2022-07-02 NYY    CLE              2

In [328]:
doubleheaders['A_Score'].value_counts()

1    2292
2      69
Name: A_Score, dtype: int64

In [329]:
doubleheaders['H_Score'].value_counts()

1    2292
2      69
Name: H_Score, dtype: int64

In [330]:
doubleheaders.reset_index(inplace = True)

In [331]:
doubleheaders = doubleheaders[doubleheaders['A_Score'] == 2]

In [332]:
doubleheaders.head()

Date A_Team H_Team  game_id  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
0 2022-08-02    COL    SDP        2        2      2            2           2   
1 2022-05-04    SDP    CLE        2        2      2            2           2   
2 2022-09-17    CIN    STL        2        2      2            2           2   
3 2022-09-08    SFG    MIL        2        2      2            2           2   
4 2022-07-02    NYY    CLE        2        2      2            2           2   

   A_Wins_L30  A_Py_Wins  A_Run_Diff  H_Score  H_Win  H_Team_Wins  H_Wins_L10  \
0           2          2           2        2      2            2           2   
1           2          2           2        2      2            2           2   
2           2          2           2        2      2            2           2   
3           2          2           2        2      2            2           2   
4           2          2           2        2      2            2           2   

   H_Wins_L30  H_Py_Wins  H_Run_Diff  
0           2          2           2  
1           2          2           2  
2           2          2           2  
3           2          2           2  
4           2          2           2

In [333]:
doubleheaders = doubleheaders[['Date', 'A_Team', 'H_Team']]

In [334]:
doubleheaders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 69 entries, 0 to 68
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    69 non-null     datetime64[ns]
 1   A_Team  69 non-null     object        
 2   H_Team  69 non-null     object        
dtypes: datetime64[ns](1), object(2)
memory usage: 2.2+ KB


In [132]:
#away_scores_df = scores_df[['Away_Team', 'Away_Score', 'Date']]

In [133]:
#away_scores_df.head()

In [134]:
#away_scores_df.info()

In [135]:
#home_scores_df = scores_df[['Home_Team', 'Home_Score', 'Date']]

In [136]:
#home_scores_df.head()

In [335]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 2429
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         2430 non-null   datetime64[ns]
 1   game_id      2430 non-null   int64         
 2   A_Team       2430 non-null   object        
 3   A_Score      2430 non-null   int64         
 4   A_Win        2430 non-null   int32         
 5   A_Team_Wins  2415 non-null   float64       
 6   A_Wins_L10   2415 non-null   float64       
 7   A_Wins_L30   2415 non-null   float64       
 8   A_Py_Wins    2415 non-null   float64       
 9   A_Run_Diff   2415 non-null   float64       
 10  H_Team       2430 non-null   object        
 11  H_Score      2430 non-null   int64         
 12  H_Win        2430 non-null   int32         
 13  H_Team_Wins  2415 non-null   float64       
 14  H_Wins_L10   2415 non-null   float64       
 15  H_Wins_L30   2415 non-null   float64       
 16  H_Py_W

# Outer merge
We'll filter the doubleheaders out of wins_21 with an outer merge.

In [336]:
#perform outer join
outer = wins_21.merge(doubleheaders, how='outer', indicator=True)



In [337]:
outer.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2430 entries, 0 to 2429
Data columns (total 19 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         2430 non-null   datetime64[ns]
 1   game_id      2430 non-null   int64         
 2   A_Team       2430 non-null   object        
 3   A_Score      2430 non-null   int64         
 4   A_Win        2430 non-null   int32         
 5   A_Team_Wins  2415 non-null   float64       
 6   A_Wins_L10   2415 non-null   float64       
 7   A_Wins_L30   2415 non-null   float64       
 8   A_Py_Wins    2415 non-null   float64       
 9   A_Run_Diff   2415 non-null   float64       
 10  H_Team       2430 non-null   object        
 11  H_Score      2430 non-null   int64         
 12  H_Win        2430 non-null   int32         
 13  H_Team_Wins  2415 non-null   float64       
 14  H_Wins_L10   2415 non-null   float64       
 15  H_Wins_L30   2415 non-null   float64       
 16  H_Py_W

There were 59 doubleheaders where a single date will have the same team twice. Therefore, 118 rows that match the doublheaders dataframe. For 2022, it's 69 and 138.

In [339]:
outer['_merge'].value_counts()

left_only     2292
both           138
right_only       0
Name: _merge, dtype: int64

In [340]:

#perform anti-join
wins_21 = outer[(outer._merge=='left_only')].drop('_merge', axis=1)

In [341]:
outer

Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
0    2022-04-07     2429    HOU        3      1          NaN         NaN   
1    2022-04-07     2430    SDP        2      0          NaN         NaN   
2    2022-04-07     2431    CIN        6      1          NaN         NaN   
3    2022-04-07     2432    MIL        4      0          NaN         NaN   
4    2022-04-07     2433    CLE        1      0          NaN         NaN   
...         ...      ...    ...      ...    ...          ...         ...   
2425 2022-10-05     4854    LAA        2      0         73.0         7.0   
2426 2022-10-05     4855    STL        3      0         93.0         5.0   
2427 2022-10-05     4856    SFG        8      1         80.0         6.0   
2428 2022-10-05     4857    DET        4      0         66.0         7.0   
2429 2022-10-05     4858    NYY        2      0         99.0         6.0   

      A_Wins_L30  A_Py_Wins  A_Run_Diff H_Team  H_Score  H_Win  H_Team_Wins  \
0            NaN        NaN         NaN    LAA        1      0          NaN   
1            NaN        NaN         NaN    ARI        4      1          NaN   
2            NaN        NaN         NaN    ATL        3      0          NaN   
3            NaN        NaN         NaN    CHC        5      1          NaN   
4            NaN        NaN         NaN    KCR        3      1          NaN   
...          ...        ...         ...    ...      ...    ...          ...   
2425        16.0       75.0       -44.0    OAK        3      1         59.0   
2426        17.0       95.0       137.0    PIT        5      1         61.0   
2427        17.0       82.0        12.0    SDP        1      0         89.0   
2428        16.0       63.0      -155.0    SEA        5      1         89.0   
2429        20.0      106.0       242.0    TEX        4      1         67.0   

      H_Wins_L10  H_Wins_L30  H_Py_Wins  H_Run_Diff     _merge  
0            NaN         NaN        NaN         NaN  left_only  
1            NaN         NaN        NaN         NaN  left_only  
2            NaN         NaN        NaN         NaN  left_only  
3            NaN         NaN        NaN         NaN  left_only  
4            NaN         NaN        NaN         NaN  left_only  
...          ...         ...        ...         ...        ...  
2425         4.0        10.0       58.0      -203.0  left_only  
2426         6.0        12.0       57.0      -228.0  left_only  
2427         6.0        17.0       86.0        52.0  left_only  
2428         6.0        16.0       88.0        66.0  left_only  
2429         2.0         9.0       77.0       -38.0  left_only  

[2430 rows x 19 columns]

In [342]:
wins_21.head()

Date  game_id A_Team  A_Score  A_Win  A_Team_Wins  A_Wins_L10  \
0 2022-04-07     2429    HOU        3      1          NaN         NaN   
1 2022-04-07     2430    SDP        2      0          NaN         NaN   
2 2022-04-07     2431    CIN        6      1          NaN         NaN   
3 2022-04-07     2432    MIL        4      0          NaN         NaN   
4 2022-04-07     2433    CLE        1      0          NaN         NaN   

   A_Wins_L30  A_Py_Wins  A_Run_Diff H_Team  H_Score  H_Win  H_Team_Wins  \
0         NaN        NaN         NaN    LAA        1      0          NaN   
1         NaN        NaN         NaN    ARI        4      1          NaN   
2         NaN        NaN         NaN    ATL        3      0          NaN   
3         NaN        NaN         NaN    CHC        5      1          NaN   
4         NaN        NaN         NaN    KCR        3      1          NaN   

   H_Wins_L10  H_Wins_L30  H_Py_Wins  H_Run_Diff  
0         NaN         NaN        NaN         NaN  
1         NaN         NaN        NaN         NaN  
2         NaN         NaN        NaN         NaN  
3         NaN         NaN        NaN         NaN  
4         NaN         NaN        NaN         NaN

In [343]:
wins_21.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2292 entries, 0 to 2429
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         2292 non-null   datetime64[ns]
 1   game_id      2292 non-null   int64         
 2   A_Team       2292 non-null   object        
 3   A_Score      2292 non-null   int64         
 4   A_Win        2292 non-null   int32         
 5   A_Team_Wins  2277 non-null   float64       
 6   A_Wins_L10   2277 non-null   float64       
 7   A_Wins_L30   2277 non-null   float64       
 8   A_Py_Wins    2277 non-null   float64       
 9   A_Run_Diff   2277 non-null   float64       
 10  H_Team       2292 non-null   object        
 11  H_Score      2292 non-null   int64         
 12  H_Win        2292 non-null   int32         
 13  H_Team_Wins  2277 non-null   float64       
 14  H_Wins_L10   2277 non-null   float64       
 15  H_Wins_L30   2277 non-null   float64       
 16  H_Py_W

# Conclusion
Now we'll save this all to a csv called XXXX_Win_Features, with XXXX indicating the year. Win_Features denotes that we've derived win variables for the dataframe. The next step will be to add starting pitcher data.

In [344]:
filepath = r'C:\Users\Owner\Sports Betting\MLB_Game_Outcome\2022_Win_Features.csv'
wins_21.to_csv(filepath)